katapult.go

```diff
@@ -92,13 +103,70 @@ func New(opts ...Opt) (*Client, error) {
 	return c, nil
 }
 
-type Client struct {
-	HTTPClient *http.Client
-	Codec      codec.Codec
+func (c *Client) DoIt(
+	ctx context.Context,
+	request *Request,
+	v interface{},
+) (*Response, error) {
+	contentType, bodyReader, err := request.bodyContent()
+	if err != nil {
+		return nil, err
+	}
 
-	APIKey    string
-	UserAgent string
-	BaseURL   *url.URL
+	u := c.BaseURL.ResolveReference(request.URL)
+	req, err := http.NewRequestWithContext(
+		ctx, request.Method, u.String(), bodyReader,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(request.Header) > 0 {
+		for k := range request.Header {
+			for _, v := range request.Header.Values(k) {
+				req.Header.Add(k, v)
+			}
+		}
+	}
+
+	if !request.NoAuth {
+		if c.APIKey == "" {
+			return nil, fmt.Errorf(
+				"%w: no API key available for: %s %s",
+				ErrRequest, request.Method, request.URL.Path,
+			)
+		}
+		req.Header.Set(
+			"Authorization",
+			fmt.Sprintf("Bearer %s", c.APIKey),
+		)
+	}
+	req.Header.Set("User-Agent", c.UserAgent)
+	req.Header.Set("Accept", "application/json")
+	if contentType != "" {
+		req.Header.Set("Content-Type", contentType)
+	}
+
+	r, err := c.HTTPClient.Do(req)
+	if err != nil {
+		return NewResponse(nil), err
+	}
+	defer r.Body.Close()
+
+	resp := NewResponse(r)
+	if resp.StatusCode/100 != 2 {
+		return c.handleResponseError(resp)
+	}
+
+	if v != nil && resp.StatusCode != 204 {
+		if w, ok := v.(io.Writer); ok {
+			_, err = io.Copy(w, r.Body)
```

Comment:
Hmm, do we want to return the response if the body fails to decode? I tend to pretend returning nil unless there's a useful meaning to the value in an error state.

Reply:
Until #64 is implemented, we kind of do. Even if everything goes to shit, a response with a status code should still be available to manually inspect. And in fact, returning `nil` as `*Response` actively causes panics in a number of places right now :(

My plan is to both the return values and the panics as part of #64 as it'll allow us to expose useful error information in a much nicer way.

Reply:
Perfect